### PR TITLE
Added Azure Stack provisioning

### PR DESCRIPTION
**Pull request by Ajay-MS**

### CODE REVIEW

File: TaskModules/powershell/Azure/SPNCreation.ps1 [hunk 1]
```diff
@@ -42,6 +133,33 @@ function Get-Password
     }
 }
 
+
+if ($environmentName.Equals($AZURESTACK_ENVIRONMENT))
```
Comment (asranja) [inline TaskModules/powershell/Azure/SPNCreation.ps1]: Should we do case-insensitive comparison ?

File: TaskModules/powershell/Azure/SPNCreation.ps1 [hunk 2]
```diff
@@ -42,6 +133,33 @@ function Get-Password
...
 
+
+if ($environmentName.Equals($AZURESTACK_ENVIRONMENT))
+{
+    if(-not $azureStackManagementURL)
+    {
+        $azureStackEnvironment = Get-AzureRmEnvironment -Name $AZURESTACK_ENVIRONMENT
+        if(-not $azureStackEnvironment)
+        {
+            Write-Output "AzureStack Enviornment is not added to AzureRmEnvironments in current PS Session. Please provide AzureStackManagementURL as argument or add AzureStack to AzureRmEnvironments manually."
```
Comment (vincentdass) [inline TaskModules/powershell/Azure/SPNCreation.ps1]: should we throw it as error?

File: TaskModules/powershell/Azure/SPNCreation.ps1 [hunk 3]
```diff
@@ -10,9 +10,100 @@ param
...
+    $AzureKeyVaultDnsSuffix="vault.$($stackdomain)".ToLowerInvariant()
+    $AzureKeyVaultServiceEndpointResourceId= $("https://vault.$stackdomain".ToLowerInvariant())
+    $StorageEndpointSuffix = ($stackdomain).ToLowerInvariant()
+
+
+    $azureStackEndpointUri = $EndpointURI.ToString() + "/metadata/endpoints?api-version=2015-01-01"
+
+    Write-Verbose "Retrieving endpoints from the $ResourceManagerEndpoint"
+
+    $endpointData = Invoke-RestMethod -Uri $azureStackEndpointUri -Method Get -ErrorAction Stop
```
Comment (vincentdass) [inline TaskModules/powershell/Azure/SPNCreation.ps1]: Do we need to catch the exception here and provide user friendly message?

File: TaskModules/powershell/Azure/SPNCreation.ps1 (hunk 2 above)
Comment (asranja) [inline TaskModules/powershell/Azure/SPNCreation.ps1]: Write-Output vs Write-Error ?

File: TaskModules/powershell/Azure/SPNCreation.ps1 [hunk 4]
```diff
@@ -10,9 +10,100 @@ param
...
+        ActiveDirectoryEndpoint                  = $activeDirectoryEndpoint
+        ActiveDirectoryServiceEndpointResourceId = $activeDirectoryServiceEndpointResourceId
+        ResourceManagerEndpoint                  = $ResourceManagerEndpoint
+        GalleryEndpoint                          = $galleryEndpoint
+        GraphEndpoint                            = $graphEndpoint
+        GraphAudience                            = $graphAudience
+        StorageEndpointSuffix                    = $StorageEndpointSuffix
+        AzureKeyVaultDnsSuffix                   = $AzureKeyVaultDnsSuffix
+        AzureKeyVaultServiceEndpointResourceId   = $AzureKeyVaultServiceEndpointResourceId
+        EnableAdfsAuthentication                 = $aadAuthorityEndpoint.TrimEnd("/").EndsWith("/adfs", [System.StringComparison]::OrdinalIgnoreCase)
```
Comment (vincentdass) [inline TaskModules/powershell/Azure/SPNCreation.ps1]: do we need to validate the case where aadAuthorityEndpoint is null?

File: TaskModules/powershell/Azure/SPNCreation.ps1 [hunk 5]
```diff
@@ -10,9 +10,100 @@ param
...
+
+    if ($endpointData)
+    {
+        $authenticationData = $endpointData.authentication;
+        if ($authenticationData)
+        {
+            $loginEndpoint = $authenticationData.loginEndpoint
+            if($loginEndpoint)
+            {
+                $aadAuthorityEndpoint = $loginEndpoint
```
Comment (vincentdass) [inline TaskModules/powershell/Azure/SPNCreation.ps1]: will loginEndpoint be null?

**Review by vincent1173 — APPROVED** [review]

File: TaskModules/powershell/Azure/SPNCreation.ps1 (hunk 3 above)
Comment (Ajay-MS) [reply TaskModules/powershell/Azure/SPNCreation.ps1]: Catching the exception in the caller of this method.

File: TaskModules/powershell/Azure/SPNCreation.ps1 [hunk 6]
```diff
@@ -10,9 +10,102 @@ param
...
+        [parameter(mandatory=$true, HelpMessage="Azure Stack environment name for use with AzureRM commandlets")]
+        [string] $Name
+    )
+
+    $EndpointURI = $EndpointURI.TrimEnd("/")
+
+    $Domain = ""
+    try {
+        $uriendpoint = [System.Uri] $EndpointURI
+        $i = $EndpointURI.IndexOf('.')
```
Comment (kmkumaran) [inline TaskModules/powershell/Azure/SPNCreation.ps1]: How safe it is to assume that anything after first dot is domain?
http://www.w3.org/Addressing/URL/url-spec.txt

File: TaskModules/powershell/Azure/SPNCreation.ps1 [hunk 7]
```diff
@@ -42,6 +135,32 @@ function Get-Password
...
+        {
+            Write-Output "ERROR: Failed to add AzureStack environment to AzureRmEnvironments";
+            throw $_.Exception
+        }
+    }
+}
+
+
 #Initialize
 $ErrorActionPreference = "Stop"
```
Comment (kmkumaran) [inline TaskModules/powershell/Azure/SPNCreation.ps1]: Your Error action preference and Verbose preference should be set before you initialize AzureStack environment

File: TaskModules/powershell/Azure/SPNCreation.ps1 (hunk 6 above)
Comment (Ajay-MS) [reply TaskModules/powershell/Azure/SPNCreation.ps1]: This URL is always in format of management.<domain>,  Same method Azure Stack team use for provisioning and our tasks use the same algorithm and has already been verified.

**Review by kmkumaran — APPROVED** [review]

**Pull request merged**